dali/python/nvidia/dali/plugin/paddle.py

```diff
@@ -307,30 +303,20 @@ def __next__(self):
                 else:
                     category_place[cat] = pd_cpu_place
 
-            if self._data_batches[i] is None:
-                pd_tensors = {}
-                for cat, lod in self.normalized_map.items():
-                    lod_tensor = fluid.core.LoDTensor()
-                    lod_tensor._set_dims(category_shapes[cat])
-                    pd_tensors[cat] = lod_tensor
-                self._data_batches[i] = pd_tensors
-            else:
-                pd_tensors = self._data_batches[i]
-
-            # Copy data from DALI Tensors to LoDTensors
+            pd_tensors = {}
             for cat, tensor in category_tensors.items():
-                if hasattr(tensor, 'shape'):  # could be tensor list
-                    assert self._dynamic_shape or \
-                        tensor.shape() == pd_tensors[cat].shape(), \
-                        ("Shapes do not match: DALI tensor has size {0}, "
-                         "but LoDTensor has size {1}".format(
-                             tensor.shape(), pd_tensors[cat].shape()))
-
-                lod_tensor = pd_tensors[cat]
+                lod_tensor = fluid.core.LoDTensor()
+                pd_tensors[cat] = lod_tensor
                 lod_tensor._set_dims(category_shapes[cat])
                 seq_len = category_lengths[cat]
                 lod_tensor.set_recursive_sequence_lengths(seq_len)
-                ptr = lod_tensor._mutable_data(category_place[cat],
+                lod_tensor._mutable_data(category_place[cat],
+                                         category_pd_type[cat])
+            data_batches[i] = pd_tensors
+
+            fluid.core._cuda_synchronize(pd_gpu_place)
+            for cat, tensor in category_tensors.items():
+                ptr = pd_tensors[cat]._mutable_data(category_place[cat],
```

Comment:
We basically call the `_mutable_data` twice, the second time after sync. Is this intended? Is this the WAR? Do we need some comment?

Reply:
The first time is to get the memory allocated. Then we sync as the WAR. Then we obtain the pointer to the memory itself (there it no other way than call `_mutable_data`)..

Reply:
Can you maybe mention the issue in a comment? For example describe that you allocate, sync as a WAR, and acquire the pointer?

Reply:
Sure.

Reply:
Done